tests/test_cart.py

```diff
@@ -1,15 +1,231 @@
 from django.core.exceptions import ObjectDoesNotExist
+from django.http import HttpRequest
 from mock import MagicMock, Mock
+from satchless.item import InsufficientStock
 
-from saleor.cart import decorators, forms, utils
+from saleor.cart import forms, utils
+from saleor.cart.decorators import (find_and_assign_cart,
+                                    get_anonymous_cart_from_token,
+                                    get_cart_from_request,
+                                    get_or_create_anonymous_cart_from_token,
+                                    get_or_create_cart_from_request,
+                                    get_or_create_user_cart, get_user_cart)
 from saleor.cart.models import Cart
-from satchless.item import InsufficientStock
+
+
+@pytest.fixture()
+def test_user(django_user_model):
+    return django_user_model.objects.get_or_create(
+        email='test@example.com', defaults={'is_active': True,
+                                            'is_staff': True,
+                                            'is_superuser': True})[0]
+
+
+@pytest.fixture()
+def opened_anonymous_cart(test_user):
+    return Cart.objects.get_or_create(user=None, status=Cart.OPEN)[0]
+
+
+@pytest.fixture()
+def cancelled_anonymous_cart(test_user):
+    return Cart.objects.get_or_create(user=None, status=Cart.CANCELED)[0]
+
+
+@pytest.fixture()
+def opened_user_cart(test_user):
+    return Cart.objects.get_or_create(user=test_user, status=Cart.OPEN)[0]
+
+
+@pytest.fixture()
+def cancelled_user_cart(test_user):
+    return Cart.objects.get_or_create(user=test_user, status=Cart.CANCELED)[0]
+
+
+def test_get_or_create_anonymous_cart_from_token(opened_anonymous_cart,
+                                                 cancelled_anonymous_cart,
+                                                 opened_user_cart,
+                                                 cancelled_user_cart):
+    queryset = Cart.objects.all()
+    carts = list(queryset)
+    cart = get_or_create_anonymous_cart_from_token(opened_anonymous_cart.token)
+    assert Cart.objects.all().count() == 4
+    assert cart == opened_anonymous_cart
+
+    # test against getting closed carts
+    cart = get_or_create_anonymous_cart_from_token(
+        cancelled_anonymous_cart.token)
+    assert Cart.objects.all().count() == 5
+    assert cart not in carts
+    assert cart.user is None
+    assert cart.status == Cart.OPEN
+    cart.delete()
+
+    # test against new token
+    cart = get_or_create_anonymous_cart_from_token(uuid4())
+    assert Cart.objects.all().count() == 5
+    assert cart not in carts
+    assert cart.user is None
+    assert cart.status == Cart.OPEN
+    cart.delete()
+
+    # test against getting cart assigned to user
+    cart = get_or_create_anonymous_cart_from_token(opened_user_cart.token)
+    assert Cart.objects.all().count() == 5
+    assert cart not in carts
+    assert cart.user is None
+    assert cart.status == Cart.OPEN
+    cart.delete()
```

Comment:
It may be useful to use `pytest.mark.parametrize` (http://doc.pytest.org/en/latest/parametrize.html) to avoid repeating logic for each case.